Unexpected error when installing new compiler

### Steps to reproduce

```console
$ git clone https://github.com/spack/spack.git ${SPACK_ROOT}
remote: Enumerating objects: 19457, done.
remote: Counting objects: 100% (19457/19457), done.
remote: Compressing objects: 100% (10112/10112), done.
remote: Total 19457 (delta 1618), reused 13918 (delta 1551), pack-reused 0 (from 0)
$ source ${SPACK_ROOT}/share/spack/setup-env.sh
$ spack compiler find
==> Added 1 new compiler to /opt/spack/etc/spack/packages.yaml
    gcc@13.3.0
==> Compilers are defined in the following files:
    /opt/spack/etc/spack/packages.yaml
$ spack install gcc@14.2.0
```

### Error message

```console
...
[+] /opt/spack/opt/spack/linux-zen4/mpc-1.3.1-zfuoq3vih6lcbiv2hhct5wdrjfrjjhq6
==> No binary for gcc-14.2.0-twh6tf5tlrhklagfwhdwsorwuldye4gw found: installing from source
==> Installing gcc-14.2.0-twh6tf5tlrhklagfwhdwsorwuldye4gw [34/34]
==> Fetching https://mirror.spack.io/_source-cache/archive/a7/a7b39bc69cbf9e25826c5a60ab26477001f7c08d85cec04bc0e29cabed6f3cc9.tar.xz
    [100%]   92.31 MB @   98.9 MB/s
Process Process-48:
Traceback (most recent call last):
  File "/opt/spack/lib/spack/spack/build_environment.py", line 1184, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/installer.py", line 2750, in build_process
    return installer.run()
           ^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/installer.py", line 2606, in run
    self.pkg.do_patch()
  File "/opt/spack/lib/spack/spack/package_base.py", line 1730, in do_patch
    self.patch()
  File "/root/.spack/package_repos/fncqgg4/repos/spack_repo/builtin/packages/gcc/package.py", line 790, in patch
    self.build_optimization_config()
  File "/root/.spack/package_repos/fncqgg4/repos/spack_repo/builtin/packages/gcc/package.py", line 827, in build_optimization_config
    flags += " " + self.get_common_target_flags(self.spec)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.spack/package_repos/fncqgg4/repos/spack_repo/builtin/packages/gcc/package.py", line 798, in get_common_target_flags
    flags = microarchitecture_flags_from_target(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/archspec.py", line 57, in microarchitecture_flags_from_target
    return target.optimization_flags(compiler.package.archspec_name(), version_number)
                                     ^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/spec.py", line 1964, in package
    assert self.concrete, "{0}: Spec.package can only be called on concrete specs".format(
           ^^^^^^^^^^^^^
AssertionError: gcc: Spec.package can only be called on concrete specs

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/spack/lib/spack/spack/build_environment.py", line 1202, in _setup_pkg_and_run
    package_context = get_package_context(tb)
                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/build_environment.py", line 1484, in get_package_context
    func = getattr(obj, tb.tb_frame.f_code.co_name, "")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/spec.py", line 1964, in package
    assert self.concrete, "{0}: Spec.package can only be called on concrete specs".format(
           ^^^^^^^^^^^^^
AssertionError: gcc: Spec.package can only be called on concrete specs
==> Error: Failed to install gcc due to InstallError: The process has stopped unexpectedly (exit 1)
```

### Information on your system

* **Spack:** 1.1.0.dev0 (https://github.com/spack/spack/commit/b187f8758227abdfc9eb349a48f8b725aa27a162)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/50ec8f1857cf2f631d20643a5049ba8b830cde29
* **Python:** 3.12.3
* **Platform:** linux-ubuntu24.04-zen4

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

--- a/repos/spack_repo/builtin/packages/gcc/package.py
+++ b/repos/spack_repo/builtin/packages/gcc/package.py
@@ -791,7 +791,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage, CompilerPackage):
 
     def get_common_target_flags(self) -> str:
         """Get the microarchitecture flags supported by both the current spec and its gcc dep.
-        For example: gcc@7 %gcc@12 target=znver3 should pick -march=znver1, since that's what
+        For example: gcc@7 target=znver3 %gcc@12 should pick -march=znver1, since that's what
         gcc@7 supports."""
         for uarch in (self.spec.target, *self.spec.target.ancestors):
             flags = microarchitecture_flags_from_target(uarch, compiler=self.spec)

--- a/repos/spack_repo/builtin/packages/gcc/package.py
+++ b/repos/spack_repo/builtin/packages/gcc/package.py
@@ -789,15 +789,12 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage, CompilerPackage):
             )
         self.build_optimization_config()
 
-    def get_common_target_flags(self, spec):
-        """Get the right (but pessimistic) architecture specific flags supported by
-        both host gcc and to-be-built gcc. For example: gcc@7 %gcc@12 target=znver3
-        should pick -march=znver1, since that's what gcc@7 supports."""
-        microarchitectures = [spec.target] + spec.target.ancestors
-        for uarch in microarchitectures:
-            flags = microarchitecture_flags_from_target(
-                uarch, compiler=Spec(f"gcc@={spec.version}")
-            )
+    def get_common_target_flags(self) -> str:
+        """Get the microarchitecture flags supported by both the current spec and its gcc dep.
+        For example: gcc@7 %gcc@12 target=znver3 should pick -march=znver1, since that's what
+        gcc@7 supports."""
+        for uarch in (self.spec.target, *self.spec.target.ancestors):
+            flags = microarchitecture_flags_from_target(uarch, compiler=self.spec)
             if flags:
                 return flags
         # no arch specific flags in common, unlikely to happen.
@@ -824,7 +821,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage, CompilerPackage):
         # Note we're not adding this for aarch64 because of
         # https://github.com/spack/spack/issues/31184
         if "+bootstrap %gcc" in self.spec and self.spec.target.family != "aarch64":
-            flags += " " + self.get_common_target_flags(self.spec)
+            flags += " " + self.get_common_target_flags()
 
         if self.spec.satisfies("+bootstrap"):
             variables = ["BOOT_CFLAGS", "CFLAGS_FOR_TARGET", "CXXFLAGS_FOR_TARGET"]

